### PR TITLE
Remove `token::Lit` from `ast::MetaItemLit`.

### PR DESCRIPTION
**Pull request by nnethercote**
Currently `ast::MetaItemLit` represents the literal kind twice. This PR removes that redundancy. Best reviewed one commit at a time.

r? @petrochenkov

### CODE REVIEW

File: compiler/rustc_ast/src/attr/mod.rs [hunk 1]
```diff
@@ -328,7 +328,9 @@ pub fn mk_name_value_item_str(ident: Ident, str: Symbol, str_span: Span) -> Meta
 }
 
 pub fn mk_name_value_item(ident: Ident, kind: LitKind, lit_span: Span) -> MetaItem {
-    let lit = MetaItemLit { token_lit: kind.to_token_lit(), kind, span: lit_span };
+    let token_lit = kind.synthesize_token_lit();
```
Comment (petrochenkov) [inline compiler/rustc_ast/src/attr/mod.rs:331]: This use of `synthesize_token_lit` is suspicious, we don't need to go through a token if we are directly constructing a `MetaItemLit`, need to check where is this `mk_name_value_item` used.

File: compiler/rustc_ast/src/attr/mod.rs [hunk 2]
```diff
@@ -408,7 +410,7 @@ pub fn mk_attr_name_value_str(
     val: Symbol,
     span: Span,
 ) -> Attribute {
-    let lit = LitKind::Str(val, StrStyle::Cooked).to_token_lit();
+    let lit = LitKind::Str(val, StrStyle::Cooked).synthesize_token_lit();
```
Comment (petrochenkov) [inline compiler/rustc_ast/src/attr/mod.rs:413]: Also suspicious for the same reason as `mk_name_value_item`.

File: compiler/rustc_expand/src/build.rs [hunk 3]
```diff
@@ -333,7 +333,7 @@ impl<'a> ExtCtxt<'a> {
     }
 
     fn expr_lit(&self, span: Span, lit_kind: ast::LitKind) -> P<ast::Expr> {
-        let token_lit = lit_kind.to_token_lit();
+        let token_lit = lit_kind.synthesize_token_lit();
```
Comment (petrochenkov) [inline compiler/rustc_expand/src/build.rs:336]: Also suspicious, if an AST builder  function produces a token lit, it shouldn't take lowered `ast::LitKind`.

File: compiler/rustc_parse/src/parser/expr.rs [hunk 4]
```diff
@@ -1551,7 +1551,7 @@ impl<'a> Parser<'a> {
                 })
             });
             consume_colon = false;
-            Ok(self.mk_expr(lo, ExprKind::Lit(lit.token_lit)))
+            Ok(self.mk_expr(lo, ExprKind::Lit(lit.as_token_lit())))
```
Comment (petrochenkov) [inline compiler/rustc_parse/src/parser/expr.rs:1554]: Also suspicious, if we are in the parser we should not use `MetaItemLit`.

File: compiler/rustc_parse/src/parser/expr.rs [hunk 5]
```diff
@@ -1817,7 +1818,7 @@ impl<'a> Parser<'a> {
     pub(super) fn parse_token_lit(&mut self) -> PResult<'a, (token::Lit, Span)> {
         self.parse_opt_token_lit()
             .ok_or(())
-            .or_else(|()| self.handle_missing_lit().map(|lit| (lit.token_lit, lit.span)))
+            .or_else(|()| self.handle_missing_lit().map(|lit| (lit.as_token_lit(), lit.span)))
```
Comment (petrochenkov) [inline compiler/rustc_parse/src/parser/expr.rs:1821]: Same as in the previous comment.

File: compiler/rustc_parse/src/parser/pat.rs [hunk 6]
```diff
@@ -420,7 +420,7 @@ impl<'a> Parser<'a> {
                 err.span_label(self_.token.span, format!("expected {}", expected));
                 err
             });
-            PatKind::Lit(self.mk_expr(lo, ExprKind::Lit(lit.token_lit)))
+            PatKind::Lit(self.mk_expr(lo, ExprKind::Lit(lit.as_token_lit())))
```
Comment (petrochenkov) [inline compiler/rustc_parse/src/parser/pat.rs:423]: Same as in the previous comment.

File: compiler/rustc_ast/src/util/literal.rs [hunk 7]
```diff
@@ -196,9 +202,32 @@ impl LitKind {
...
+        Ok(MetaItemLit {
+            symbol: token_lit.symbol,
+            suffix: token_lit.suffix,
+            kind: LitKind::from_token_lit(token_lit)?,
+            span,
+        })
+    }
+
+    /// Cheaply converts a meta item literal into a token literal.
+    pub fn as_token_lit(&self) -> token::Lit {
```
Comment (petrochenkov) [inline compiler/rustc_ast/src/util/literal.rs:216]: Basically, the only non-suspicious uses of `synthesize_token_lit` and this function are AST/HIR pretty-printing (including rustfmt), and recovering a token from `ExprKind::IncludedBytes`.

File: compiler/rustc_ast/src/attr/mod.rs (hunk 1 above)
Comment (nnethercote) [reply compiler/rustc_ast/src/attr/mod.rs]: Removed.

File: compiler/rustc_ast/src/attr/mod.rs (hunk 2 above)
Comment (nnethercote) [reply compiler/rustc_ast/src/attr/mod.rs:398]: Removed.

File: compiler/rustc_expand/src/build.rs (hunk 3 above)
Comment (nnethercote) [reply compiler/rustc_expand/src/build.rs]: Removed.

File: compiler/rustc_parse/src/parser/expr.rs (hunk 4 above)
Comment (nnethercote) [reply compiler/rustc_parse/src/parser/expr.rs]: Fixed.

File: compiler/rustc_parse/src/parser/expr.rs (hunk 5 above)
Comment (nnethercote) [reply compiler/rustc_parse/src/parser/expr.rs]: Fixed.

File: compiler/rustc_parse/src/parser/pat.rs (hunk 6 above)
Comment (nnethercote) [reply compiler/rustc_parse/src/parser/pat.rs]: Fixed.

File: compiler/rustc_ast/src/util/literal.rs (hunk 7 above)
Comment (nnethercote) [reply compiler/rustc_ast/src/util/literal.rs:233]: I ended up removing `synthesize_token_lit`.

**Comment by nnethercote** [comment]
I have added five new commits addressing the review comments.

**Comment by petrochenkov** [comment]
The `src/test\run-make\coverage-reports` failure may be legitimate.
@rustbot author

**Comment by nnethercote** [comment]
@richkadel: this PR makes a few changes to the AST. Nothing especially notable, and the kind of thing I've done many times before. When merging there was a single coverage test failure, but only on Windows:
```
running 76 tests
test [run-make] src/test\run-make\coverage-reports ... FAILED
<...many successful and ignored test result lines omitted...>

failures:

---- [run-make] src/test\run-make\coverage-reports stdout ----

error: make failed
status: exit code: 2
command: "make"
--- stdout -------------------------------
make[1]: Entering directory '/d/a/rust/rust/src/test/run-make/coverage-reports'
make[1]: Leaving directory '/d/a/rust/rust/src/test/run-make/coverage-reports'
------------------------------------------
--- stderr -------------------------------
Makefile:83: clear_expected_if_blessed: No such file or directory
------------------------------------------

failures:
    [run-make] src/test\run-make\coverage-reports
```
Any idea what the problem might be? I have previously seen coverage failures with quite descriptive failure messages, but this one just has the `clear_expected_if_blessed` line. I admit I don't even understand what that part of the Makefile is doing.

**Comment by nnethercote** [comment]
[This Zulip thread](https://rust-lang.zulipchat.com/#narrow/stream/242791-t-infra/topic/run-make.2Fcoverage-reports) indicates that this `coverage-reports` failure is happening to other people as well.

**Comment by richkadel** [comment]
Good to know it's unrelated to your change. `make` is really not good but it was the only tool available other than writing something from scratch. I haven't been working in the Rust codebase in a while so hopefully someone with more recent/relevant knowledge will figure out and fix the Makefile compatibility issues.

**Comment by nnethercote** [comment]
#105477 has hopefully fixed the `coverage-reports` failure.

@bors r=petrochenkov

**Comment by nnethercote** [comment]
There is a single trivial regression, which may or may not be real.

@rustbot label: +perf-regression-triaged

**Pull request merged**